Exercise: Component settings (#19)
Bixal/storybook-rapid-prototyping: stories/exercises/06_exercise.mdx

Author: This exercise was built off of exercise 5 so the other files can be ignored until exercise 5 is merged

Reviewer: #13 has been merged.

1. To get started, lets try updating the font to a serif font. In `uswds-theme.scss`, add the following line to your settings map:\
    ```scss
    $theme-banner-font-family: "serif",
    ```
1. Easy, right? Let's add a bit more color to our banner. Let's see what would happen if we brought back our accent color:
    ```scss
    $theme-banner-background-color: "accent-cool"
    ```
    The banner updated, but if you have terminal window open you might have noticed a bunch of contrast warnings pop up. USWDS does our best to ensure appropriate color contrast throughout the design system by setting good defaults. Once the default is changed, we check again where we can to ensure compliance and even try to implement a fallback color if it's not.

    If the automated checks determine neither color meets contrast requirements, we'll allow you to override but give you a warning. If we inspect this warning, we learn that the link color and it's fallback do not meet color contrast minimums. Luckily, there's a setting for that!

> 1. Find the setting for the banner link color and a token that we defined in exercise 5 to prevent the contrast error.

Author: **question:** How do we feel about the more "informative" guidance here? In previous exercises, we didn't include much but I could see value in the long term if we include it similar to tutorial.

Reviewer: That type of guidance can be useful. We might want to prefix it with `**note:**` So that users can quickly see.

Author: Added this in [235b685](https://github.com/Bixal/storybook-rapid-prototyping/pull/19/commits/235b685d6f27e3033b4ce4803fc309cb42cba4f4)

@mejiaj  let me know what you think of the placement! I think this feels like the appropriate place for it.

    **Tip:** The color should be darker than the `accent-cool` token.
1. Save the theme file. If you no longer see a warning in your compile log, visit the landing page and confirm the link color is a good match for it's background.

> 1. Next, lets use a header [component setting](https://designsystem.digital.gov/components/header/#implementation) to change the breakpoint that it collapses into the mobile nav. Since we don't have many navigation links, lets decrease the screen size needed to trigger the non-mobile header.

Author: **Thought:** Exercise 6 feels a bit longer than the other exercises. If we want to prune, we do a different `min-width` update later in this exercise and could get rid of the header setting

# 06. Customize components using settings.

In the last exercise, we familiarized ourselves with the USWDS theme settings. In this section, we'll take that a step further by using component specific settings.

Each USWDS component with settings will have this section. It is a snapshot of what customizations are possible without needing to write custom styles.

**Tip:** The [Component settings](https://designsystem.digital.gov/documentation/settings/#component-settings-2) section of the Settings page provides a comprehensive list of all available component settings, their default values, and what type of token or value they take!

> 1. Visit the [banner component](https://designsystem.digital.gov/components/banner/) page. Find the "Banner settings" section.

Reviewer: ```suggestion
1. Look over the available banner theme settings on the "Banner settings" section of the [component page](https://designsystem.digital.gov/components/banner/).
```

1. To get started, lets try updating the font to a serif font. In `uswds-theme.scss`, add the following line to your settings map:
    ```scss
    $theme-banner-font-family: "serif",
    ```
1. Easy, right? Let's add a bit more color to our banner. Let's see what would happen if we brought back our accent color:
    ```scss
    $theme-banner-background-color: "accent-cool"
    ```

>     The banner background color updated, but if you take a close at the banner link the contrast doesn't look great. If you take a look at the script running in the terminal window you'll notice a bunch of color contrast warnings have popped up.

Reviewer: Consider adding the message incase terminal isn't up.

```
   ```sh
       Warning: [Banner] Neither the specified preferred link color token (`base-lighter`) nor the fallback link color token (`primary`) have AA contrast on a `accent-cool` background. Using `primary`. Please check your source code and project settings.
       node_modules/@uswds/uswds/packages/uswds-core/src/styles/functions/color/get-color-token-from-bg.scss 65:7   get-color-token-from-bg()
    ```
```

Author: Resolved in [0b26483](https://github.com/Bixal/storybook-rapid-prototyping/pull/19/commits/0b26483ce4bd6fcdb6ba701474c29e69315df43c)

    USWDS does its best to ensure appropriate color contrast throughout the design system by setting good defaults. If a value is changed, it will check again where possible to ensure compliance. If the new color does not meet compliance standards, a fallback will be used.

>     If the automated checks determine neither color meets contrast requirements, we'll allow you to override but give you a warning. If we inspect this warning, we learn that the link color and it's fallback do not meet color contrast minimums.

Reviewer: ```suggestion
    If the automated checks determine neither color meets contrast requirements, we'll allow you to override with a contrast warning. If we inspect this warning, we learn that the link color and it's fallback do not meet color contrast minimums.
```

Author: Resolved in [30ff7e3](https://github.com/Bixal/storybook-rapid-prototyping/pull/19/commits/30ff7e3a62135fe81aea375663e1d8c45ab6c99c)

# 06. Customize components using settings.

> In the last exercise, we familiarized ourselves with the USWDS theme settings. In this section, we'll take that a step further by using component specific settings.

Author: Throughout this exercise, I use guidance verbiage such as:

> we familiarized...
> If we inspect...
> we can change which token...

How do you feel about this @mejiaj ? Should I avoid the collective we when referring to the user?

Reviewer: I don't see any harm in it. We can re-visit as this grows. 😄

> Each USWDS component with settings will have this section. It is a snapshot of what customizations are possible without needing to write custom styles.
> 
> **Tip:** The [Component settings](https://designsystem.digital.gov/documentation/settings/#component-settings-2) section of the Settings page provides a comprehensive list of all available component settings, their default values, and what type of token or value they take!
> 
> 1. Look over the available banner theme settings on the "Banner settings" section of the [component page](https://designsystem.digital.gov/components/banner/).
> 

Author: After moving the guidance around, this "each USWDS component with settings will have this section" line doesn't make sense. 

**question:** What do you think about moving line 5 or even 5 + 7 under line 9?

```suggestion
**Tip:** The [Component settings](https://designsystem.digital.gov/documentation/settings/#component-settings-2) section of the Settings page provides a comprehensive list of all available component settings, their default values, and what type of token or value they take!

1. Look over the available banner theme settings on the "Banner settings" section of the [component page](https://designsystem.digital.gov/components/banner/).

Each USWDS component with settings will have this section. It is a snapshot of what customizations are possible without needing to write custom styles.
```

Reviewer: I think it works for now, but we can revisit.

    ```

    **Note:** USWDS does its best to ensure appropriate color contrast throughout the design system by setting good defaults. If a value is changed, it will check again where possible to ensure compliance. If the new color does not meet compliance standards, a fallback will be used.

    If the automated checks determine neither color meets contrast requirements, USWDS will accept the override with a contrast warning. If we inspect this warning, we learn that the link color and it's fallback do not meet color contrast minimums.

    Luckily, there's a setting for that!
1. Find the setting for the banner link color and set it to the `primary-darker` theme token.

    **Tip:** USWDS color tokens use a 100-point [grading system](https://designsystem.digital.gov/design-tokens/color/overview/#magic-number-2) that ensures values across different color families are equal in lightness.

    A difference of 50+ results in AA color contrast compliance between any two color family tokens! This is known as the [_magic number_](https://designsystem.digital.gov/design-tokens/color/overview/#magic-number-2).

>     Since the `accent-cool` token has a grade of `30` and the `primary-darker` token has a grade of `80`, there is a different of 50. This means the two color tokens will have appropriate color contrast from each other.

Author: **note:** new color grade guidance additions
